src/main/kotlin/org/rust/lang/format/RsFormatMacroInjector.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.format
+
+import com.intellij.lang.injection.MultiHostInjector
+import com.intellij.lang.injection.MultiHostRegistrar
+import com.intellij.psi.PsiElement
+import org.rust.ide.injected.RsSimpleMultiLineEscaper
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.impl.RsLitExprImpl
+
+class RsFormatMacroInjector : MultiHostInjector {
+    override fun elementsToInjectIn(): List<Class<out PsiElement>> = listOf(RsLitExpr::class.java)
+
+    override fun getLanguagesToInject(registrar: MultiHostRegistrar, context: PsiElement) {
+        if (context.isValid
+            && context is RsLitExpr
+            && context.kind is RsLiteralKind.String) {
+
+            val macroCallExpr = (context.parent?.parent as? RsFormatMacroArgument)?.parent as? RsMacroCall
+                ?: return
+
+            if (macroCallExpr.identifier?.text in listOf(
+                    "println",
+                    "print",
+                    "writeln",
+                    "write",
+                    "eprintln",
+                    "eprint",
+                    "format_args"
+                ))
+                return
+
+            val rangeInHost = (context.kind!! as RsLiteralKind.String).offsets.value ?: return
+            (context as? RsLitExprImpl)?.overridenEscaper = RsSimpleMultiLineEscaper(context)
```

Comment:
But then there will be no escapes highlighting

Reply:
This is a discussion question. I think that we should separate rust escape codes from format pattern lexing. I have these arguments:
1. It's a good division of responsibility. Imagine that Rust escape codes are changed. In this case, you have to implement changes twice. Bugs also will be duplicated. Or there will be different bugs (as I see, you missed to implement `\x` escape code, for example).
2. If it is a language injection, it does not have to know about Rust (another language) escape codes. Imagine a project where "Rust Format pattern" used inside some TOML file, for example, for i18n purposes. TOML has different escape codes than Rust, so the user will not be able to inject "Rust Format pattern" lang into such TOML literals.
3. Nobody uses injections this way across JetBrains infrastructure. This is a Kotlin: 
![image](https://user-images.githubusercontent.com/3221931/59436856-14fcdd80-8df9-11e9-82dc-f442a4e0b256.png)
The first line is RegExp injection into Kotlin string literal, and the second line is a literal without injection. There are no escape codes highlighting in RegExp injection! `\n` is specifically lexed (see `RegExpTT.CTRL_CHARACTER`), but the lexer doesn't know written as "\n" in the origin literal. It knows only that it is LF char (which can be written just as a linebreak in a literal).

So are we. This is Rust:
![image](https://user-images.githubusercontent.com/3221931/59438894-fdbfef00-8dfc-11e9-898b-77fae241394f.png)
The first line is RegExp injection into Rust string literal, and the second line is a literal without injection. No escape codes highlighting inside injected code.

CLion highlighting is very interesting:
![image](https://user-images.githubusercontent.com/3221931/59439176-73c45600-8dfd-11e9-816a-2d4da70b91b1.png)
It's not a language injection because I see the available intention to injection something:
![image](https://user-images.githubusercontent.com/3221931/59439283-a79f7b80-8dfd-11e9-84f2-f49d2f3059da.png)
(It should be unavailable if it is already injected). Soo looks like in CLion this feature was made without language injection. I'll try to investigate it more deeply and assess whether we should do the same.

But anyway, if it is language injection, I think such hacks should not be used and we should come to terms with the fact that escape codes aren't highlighted in injections.

Reply:
Also i have just checked and following code successfully works
```
println!(" \u{007b}}",1)
```
So looks like you are right and it is a bad idea.

I can't look into clion myself, so i will wait for you to check it out

Reply:
I found that CLion uses regular `Annotator` to highlight format pattern specifiers and it parses them with, umm, regexps, so I don't think this is a good role model =/

Reply:
> so I don't think this is a good role model =/

By '"this" do you mean language injection or annotator approach?

Reply:
I meant regexp approach =D
It's possible to use a lexer (or a parser) inside an annotator to locate highlightable ranges. Such approach __can__ be used, but it's necessary to understand trade-offs... which are not clear to me.
Looks like we anyway will be forced to use some kind of pattern parser in annotator to highlight errors (e.g. parameters quantity `println!("{}", 1, 2)`). I don't know what option is better... I'll try to think of something a bit later.

Reply:
I discussed with @undin and found that original issue #2256 wasn't supposed to be solved with language injections =/ Now I think it's reasonable to use Annotator approach, but I think that string escaping should not be re-implement. I see a rough plan so:

1. Make something like `RsFormatPatternAnnotator`
2. Locate string literals that are format patterns
3. Use `RsLitExpr.kind` to take offsets and literal text
4. Use `parseRustStringCharacters` to take unescaped text and offsets
5. Parse text with our parser
6. Locate elements that should be highlighted in the parsed PSI
7. Map start and end offset of annotation with an array we got from `parseRustStringCharacters`

Then the same (single-time) parsed pattern could be used to highlight illegal format macro usage (too many or too little arguments, wrong types, etc). This is unachievable with language injections. In this case we should use a name like `RsFormatMacroAnnotator` instead of `RsFormatPatternAnnotator`

Reply:
It looks like you are suggesting to use standalone PSI parser and lexer. But to me it seems that there is no reason to use PSI parser and lexer with handful of files instead of one line regex to parse such simple syntax especially considering that we are not going to use any PSI infrastructure. Also it will create a lot of useless objects. Also I will have to manually handle parser errors.

Reply:
I'm not much familiar with regexes, so for me it's simpler to write a standalone parser =))0)
Note that it would be cool to provide a full analysis of format macros in the future (), so I think it would be handy to have PSI for format pattern to simplify analyzing of complex patterns like `println!("{number:>0width$}", number=1, width=6);`. Maybe regexes are enough, I don't sure. Maybe I just don't like regexes